src/masternodes/mn_checks.cpp

```diff
@@ -3083,6 +3097,93 @@ class CCustomTxApplyVisitor : public CCustomTxVisitor
         return mnview.AddBalance(obj.to, obj.amount);
     }
 
+    Res operator()(const CPaybackWithCollateralMessage& obj) const {
+        auto res = CheckCustomTx();
+        if (!res)
+            return res;
+
+        const auto attributes = mnview.GetAttributes();
+        if (!attributes)
+            return Res::Err("Attributes unavailable");
+
+        auto dUsdToken = mnview.GetToken("DUSD");
+        if (!dUsdToken)
+            return Res::Err("Cannot find token DUSD");
+
+        CDataStructureV0 activeKey{AttributeTypes::Token, dUsdToken->first.v, TokenKeys::LoanPaybackCollateral};
+        if (!attributes->GetValue(activeKey, false))
+            return Res::Err("Payback of DUSD loan with collateral is not currently active");
+
+        // vault exists
+        auto vault = mnview.GetVault(obj.vaultId);
+        if (!vault)
+            return Res::Err("Vault <%s> not found", obj.vaultId.GetHex());
+
+        // vault under liquidation
+        if (vault->isUnderLiquidation)
+            return Res::Err("Cannot payback vault with collateral while vault's under liquidation");
+
+        // owner auth
+        if (!HasAuth(vault->ownerAddress))
+            return Res::Err("tx must have at least one input from token owner");
+
+        if (!IsVaultPriceValid(mnview, obj.vaultId, height))
+            return Res::Err("Cannot payback vault with collateral while any of the asset's price is invalid");
```

Comment:
why not? we can always payback loans with same coin, cause no oracle price needed.

Reply:
Let me answer this on the main comments for better visibility later.